gst/nnstreamer/tensor_query/tensor_query_common.c

```diff
@@ -89,3 +140,397 @@ nnstreamer_query_connect (uint64_t id, const char *ip, uint32_t port,
+  GError *err;
+  if (!sdata)
+    return;
+
+  switch (sdata->protocol) {
+    case _TENSOR_QUERY_PROTOCOL_TCP:
+      if (sdata->server_socket) {
+        if (!g_socket_close (sdata->server_socket, &err)) {
+          nns_logw ("Failed to close socket: %s", err->message);
+          g_error_free (err);
+        }
+        g_object_unref (sdata->server_socket);
+      }
+      g_object_unref (sdata->cancellable);
+      break;
+    default:
+      /* NYI */
+      nns_logw ("Invalid protocol");
+      break;
+  }
+  g_free (sdata->server_host);
+  g_free (sdata);
+}
+
+/**
+ * @brief set server_data params and setup server
+ * @return 0 if OK, negative value if error 
+ */
+int
+nnstreamer_query_server_data_setup (void *server_data,
+    TensorQueryProtocol protocol, const char *host, uint32_t port)
+{
+  TensorQueryServerData *sdata = (TensorQueryServerData *) server_data;
+  if (!sdata)
+    return -EINVAL;
+
+  sdata->protocol = protocol;
+  sdata->server_host = g_strdup (host);
+  sdata->server_port = port;
+
+  switch (protocol) {
+    case _TENSOR_QUERY_PROTOCOL_TCP:
+      sdata->server_socket = query_establish_socket (host, port);
+      if (!sdata->server_socket) {
+        nns_loge ("Failed establish socket");
+        return -EIO;
+      }
+      break;
+    default:
+      /* NYI */
+      return -ENOSYS;
+  }
+
+  /* init table */
+  if (NULL == table) {
+    G_LOCK (splock);
+    table = g_hash_table_new_full (g_int64_hash, g_int64_equal, g_free, g_free);
+    // g_hash_table_unref (table);
+    G_UNLOCK (splock);
+  }
+
+  return 0;
+}
+
+/**
+ * @brief accept client connection
+ * @param server_data TensorQueryServerData
+ * @return int64_t client_id
+ */
+int64_t
+nnstreamer_query_server_accept (void *server_data)
+{
+  TensorQueryServerData *sdata = (TensorQueryServerData *) server_data;
+  GError *err;
+  int64_t *key;
+  gchar *client_host;
+  guint32 client_port;
+
+  switch (sdata->protocol) {
+    case _TENSOR_QUERY_PROTOCOL_TCP:
+    {
+      GSocket *client_socket;
+      GSocketAddress *saddr;
+      TensorQueryConnection *conn;
+      client_socket = g_socket_accept (sdata->server_socket,
+          sdata->cancellable, &err);
+      if (!client_socket) {
+        nns_loge ("Failed to accept client: %s", err->message);
+        g_error_free (err);
+        return 0;
+      }
+      saddr = g_socket_get_remote_address (client_socket, &err);
+      if (!saddr) {
+        nns_loge ("Failed to get client address: %s", err->message);
+        g_error_free (err);
+        return 0;
+      }
+
+      conn = g_new0 (TensorQueryConnection, 1);
+      conn->protocol = sdata->protocol;
+      client_host =
+          g_inet_address_to_string (g_inet_socket_address_get_address (
+              (GInetSocketAddress *) saddr));
+      client_port =
+          g_inet_socket_address_get_port ((GInetSocketAddress *) saddr);
+      conn->socket = client_socket;
+      conn->cancellable = g_cancellable_new ();
+
+      key =
```

Comment:
resource leak?
Um.. Who and How do we destroy the hash table?

Reply:
It's to be handled by sink. Sink will reuse the connection value, table key can be destroyed once sink has done with sending response

Reply:
The key generated from the source should be managed from the source.
Let's release the key when connection is closed.

Reply:
I see. I think calling `nnstreamer_query_close` (L250) can do.. can it??